Add Onboarding Wizard page

yscik opened the pull request:
#### Pending questions:

* **Fullscreen Onboarding Wizard:** 
   Does this page need to be fullscreen?
   Woo has a component that supports this, but it is not really exposed to be as an externally usable component right now. It works by hiding all the menus with CSS, which we can copy. 
* **Dismissing**:
   Can the Onboarding Wizard be closed any time without going through?  What pages do we want to redirect to it until the user finished/dismissed it?

* **Redirect to Wizard after plugin activation:**
  Added a new card for this as it is not trivial, and also partially depends on the previous question. 

#### Changes proposed in this Pull Request:

* Add Onboarding class
* Add a Sensei submenu page for onboarding. The menu item should be hidden later, but now it makes testing easier.
* Add JSX file that renders the React-based on that page
* Add components from Wordpress and Woocommerce component libraries, with styles

#### Testing instructions:

* Requires `change/build-webpack` branch
  * Merge that branch into this
  * Run `npm install`
  * Add `'onboarding/onboarding.jsx'` to `webpack.config.js`
  * Run `npm start` or `npm run build` 
* Open page in wp-admin under Sensei LMS -> Onboarding
* Page should have a card, heading and a button, properly styled.

renatho commented on assets/onboarding/onboarding.jsx:
```diff
@@ -0,0 +1,19 @@
+	);
+}
+
+wp.element.render(
```
To keep the consistency, we could import the render from the element package. WDYT? `import { render } from '@wordpress/element';`

yscik replied:
Good idea, changed it to this.

renatho commented on includes/admin/class-sensei-onboarding.php:
```diff
@@ -0,0 +1,61 @@
+			add_submenu_page( 'sensei', __( 'Onboarding', 'sensei-lms' ), __( 'Onboarding', 'sensei-lms' ), 'manage_sensei', $this->page_slug, array ( $this, 'onboarding_page' ) );
+		}
+
+	}
```
There are some indenting troubles in this file and some PHPDocs missing.

yscik replied:
Fixed in 46e2cdf hopefully.

renatho commented on includes/class-sensei.php:
```diff
@@ -184,6 +184,10 @@ class Sensei_Main {
 	 * @var Sensei_Enrolment_Job_Scheduler
 	 */
 	private $enrolment_scheduler;
+	/**
```
There is missing a break line here.

yscik replied:
46e2cdf

donnapep commented:
> * Does this page need to be fullscreen?

Ideally yes, but would depend on how much effort would be required to implement it.

> * Can the Onboarding Wizard be closed any time without going through?  What pages do we want to redirect to it until the user finished/dismissed it?

If it's full-screen, they shouldn't be able to close it. The only place they can close it currently is on [the _Welcome_ step](https://github.com/Automattic/sensei/issues/3037).

renatho commented:
WDYT about to change temporarily the target of this PR to `change/build-webpack`? So for review it'll show only this branch changes 😊

yscik commented:
> WDYT about to change temporarily the target of this PR to change/build-webpack? So for review it'll show only this branch changes 😊

It was supposed to be `next`, fixed that, and removed the accidental webpack merge pushed. Should be in a reviewable state now :)

Also added fullscreen support and a place for the stepper. This is everything for this PR, apologies for the mess.

yscik commented on assets/onboarding/fullscreen.jsx:
```diff
@@ -0,0 +1,20 @@
+
+import './fullscreen.scss';
+
+export function FullScreen( { children } ) {
```
Good idea! 7bfb324

yscik commented on assets/onboarding/fullscreen.scss:
```diff
@@ -0,0 +1,60 @@
+    display: block
+  }
+
+  * {
```
This is mostly for our own CSS and things we put on this page, so the only thing it can affect that is outside our control is the Wordpress and Woocommerce components we use. Some of those already have `border-box` set up in their component styles, and both Gutenberg and Woocommerce-admin's CSS resets include it, so this is their expected environment.

renatho replied:
Hum! I got! In this case, what do you think to move it to the `assets/onboarding/onboarding.scss` under the `.sensei-onboarding` scope? Does it make sense to you? Or could it not cover a case?

yscik replied:
I was thinking about this too, but it led to a bit of refactoring. I added a new `sensei-onboarding__page` class to the body, and extended the `useFullscreen` hook to support this. (Might need a better name)

We need this because `Modal` components are rendered outside the app and attached directly to the body, so anything scoped for our original `sensei-onboarding` container would not affect them.

renatho replied:
Hum! I got your point! So I think we can go ahead with this and if we see an error in the future we can investigate some workaround based on the error.

And I liked your approach to make to hook dynamic. =)

yscik commented on includes/admin/class-sensei-onboarding.php:
```diff
@@ -0,0 +1,100 @@
+				add_filter(
+					'admin_body_class',
+					function( $classes ) {
+						$classes .= ' sensei-wp-admin-fullscreen ';
```
That unfortunately leads to a flash when opening this page, where the wp-admin elements are visible until our component is rendered. It's actually setting the classes on the frontend that's not really needed right now, it's just there for some more robustness.

renatho replied:
Hum! I got! So good approach! =)

renatho commented on assets/onboarding/use-fullscreen.jsx:
```diff
@@ -1,18 +1,27 @@
+ * @param {string[]} bodyClasses Additional classes to be set.
+ */
+export function useFullScreen( bodyClasses = [] ) {
+	bodyClasses.push( 'sensei-wp-admin-fullscreen' );
```
I think we could keep the attribute immutable here. Maybe something like `const classes = [ ...bodyClasses, 'sensei-wp-admin-fullscreen' ]`

yscik replied:
Yup, 9c4ecdc.

renatho commented on assets/onboarding/use-fullscreen.jsx:
```diff
@@ -1,18 +1,27 @@
 	}
 
-	useEffect( setupGlobalStyles, [] );
+	useLayoutEffect( setupGlobalStyles, [] );
```
And here I think would be good to add the `bodyClasses` to the dependencies. BTW, we could add these eslint rules: https://reactjs.org/docs/hooks-rules.html#eslint-plugin WDYT?

yscik replied:
Good idea. Although calling `useFullscreen` with new classes wouldn't exactly behave as expected  right now, as the old ones wouldn't be removed. Added in 9c4ecdc.
I took a look and the React and react-hooks Lint rules are already in there, via `@wordpress/eslint-plugin/recommended`.

renatho replied:
Oh, excellent! \o/
I'm sorry that I didn't check it before to comment.

yscik commented:
I discovered some minor issue with importing CSS from JS entry points: this leads to webpack generating a `[filename].js` and `[filename].css`, but only one `[filename].asset.php`, so we cannot use separate version hash and dependency list for the JS and CSS files. 
The files are now renamed to `index.js` and `style.scss`, and additional styles are imported from the latter, which makes things a bit cleaner.

renatho commented on assets/onboarding/use-fullscreen.jsx:
```diff
@@ -1,16 +1,27 @@
 	}
 
-	useEffect( setupGlobalStyles, [] );
+	useLayoutEffect( setupGlobalStyles, [] );
```
Out of curiosity, did you test whether this solves the blink issue? Maybe it could be the reason. So we could remove that PHP workaround :)

yscik replied:
It solves a different blink issue :) `html.wp-toolbar` adds some top margin to the page, and I couldn't find a filter to remove that toolbar class wp-admin, so useFullscreen does this. But useEffect would run after the component is rendered, so there was a (much less noticeable) jump of the header upon load.
